src/Microsoft.ML.TimeSeries/SequentialTransformerBase.cs

```diff
@@ -20,13 +23,13 @@ namespace Microsoft.ML.Runtime.TimeSeriesProcessing
     /// <typeparam name="TInput">The input type of the sequential processing.</typeparam>
     /// <typeparam name="TOutput">The dst type of the sequential processing.</typeparam>
     /// <typeparam name="TState">The state type of the sequential processing. Must be a class inherited from StateBase </typeparam>
-    public abstract class SequentialTransformerBase<TInput, TOutput, TState> : ITransformer, ICanSaveModel
+    public abstract class SequentialTransformerBase<TInput, TOutput, TState> : IStatefulTransformer, ICanSaveModel
        where TState : SequentialTransformerBase<TInput, TOutput, TState>.StateBase, new()
     {
         /// <summary>
         /// The base class for encapsulating the State object for sequential processing. This class implements a windowed buffer.
         /// </summary>
-        public abstract class StateBase
+        public abstract class StateBase : ICanSaveModel, ICloneable
```

Comment:
>ICanSaveModel [](start = 42, length = 13)

Please don't use `ICanSaveModel`. At the risk of speaking tautologically, `ICanSaveModel` is for objects that save themselves as models. You're not doing that; these things that *contain* these state objects are writing themselves as models, but this object is not -- it is just writing and saving itself to a binary loader. You're just using the `BinaryWriter` and on the flipside using the `BinaryReader`. You could just as easily have done this by having a `private protected` or `internal` virtual or abstract method on this class, with an `internal`. And on the flipside you would have had a nice, hidden, internal or private protected constructor.

Reply:
Generally, unless there's some sort of "loadable class" at the other end that knows how to load this as a model, it shouldn't implement `ICanSaveModel`. Go to every implementation of `ICanSaveModel.Save`. You will see something like this at the head of it:

[https://github.com/dotnet/machinelearning/blob/08947ef1a8d65e06e1149dff8a6d13c6133c4426/src/Microsoft.ML.Data/Prediction/Calibrator.cs#L371-L372](https://github.com/dotnet/machinelearning/blob/08947ef1a8d65e06e1149dff8a6d13c6133c4426/src/Microsoft.ML.Data/Prediction/Calibrator.cs#L371-L372)

You will also see a corresponding entry, at the head of the file, something like this:

[https://github.com/dotnet/machinelearning/blob/08947ef1a8d65e06e1149dff8a6d13c6133c4426/src/Microsoft.ML.Data/Prediction/Calibrator.cs#L46-L48](https://github.com/dotnet/machinelearning/blob/08947ef1a8d65e06e1149dff8a6d13c6133c4426/src/Microsoft.ML.Data/Prediction/Calibrator.cs#L46-L48)

This is how we save and load *models*. Not just raw bitstreams, it's not for that, but real honest-to-God models.

---
In reply to: [236479112](https://github.com/dotnet/machinelearning/pull/1727#discussion_r236479112) [](ancestors = 236479112)

Reply:
This is actually something I observe in both these implementations of an interface... you picked two interfaces *seemingly at random* that "kinda sorta" did what you wanted, even though nothing at all relies (or should rely) on them implementing either interface, and just sort of wrote methods. When thinking "should this implement some random interface," it usually pays to have some sort of scenario in mind that actually requires *that interface*, which neither of these did.

---
In reply to: [236481050](https://github.com/dotnet/machinelearning/pull/1727#discussion_r236481050) [](ancestors = 236481050,236479112)